In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import matplotlib
plt.rcParams['axes.labelsize'] = 14
plt.rcParams['xtick.labelsize'] = 12
plt.rcParams['ytick.labelsize'] = 12

# KOSPI200 불러오기 & 데이터셋 생성

In [4]:
import pickle
with open('raw_data_20180811.pickle','rb') as handle:

    raw_data = pickle.load(handle)
    
kospi200 = pd.read_excel("C:\\Users\\axasd\\Downloads\\상장법인목록.xlsx")
kospi200 = kospi200[kospi200.columns[:2]]
kospi200["종목코드"] = kospi200["종목코드"].map('{:06d}'.format)

code_dict = {
    i : j for i, j in zip(kospi200["회사명"], kospi200["종목코드"])
}

In [5]:
for i in raw_data.keys():
    raw_data[i] = raw_data[i].assign(sum_close_5 = raw_data[i].Close.rolling(window =5).sum())
    raw_data[i] = raw_data[i].assign(sum_close_7 = raw_data[i].Close.rolling(window =7).sum())
    raw_data[i] = raw_data[i].assign(sum_close_9 = raw_data[i].Close.rolling(window =9).sum())
    raw_data[i] = raw_data[i].assign(ma_3 = raw_data[i].Close.rolling(window = 3).mean())
    raw_data[i] = raw_data[i].assign(ma_5 = raw_data[i].Close.rolling(window = 5).mean())
    raw_data[i] = raw_data[i].assign(ma_10 = raw_data[i].Close.rolling(window = 10).mean())
    raw_data[i] = raw_data[i].assign(ma_15 = raw_data[i].Close.rolling(window = 15).mean())

for i in raw_data.keys():
    raw_data[i] = raw_data[i].assign(pct_change5 = raw_data[i].sum_close_5.pct_change().fillna(0))
    raw_data[i] = raw_data[i].assign(pct_change7 = raw_data[i].sum_close_7.pct_change().fillna(0))
    raw_data[i] = raw_data[i].assign(pct_change9 = raw_data[i].sum_close_9.pct_change().fillna(0))
    raw_data[i] = raw_data[i].assign(sum_5_Y = np.where(raw_data[i]["pct_change5"].fillna(0)>=0, 1,0))
    raw_data[i] = raw_data[i].assign(sum_7_Y = np.where(raw_data[i]["pct_change7"].fillna(0)>=0, 1,0))
    raw_data[i] = raw_data[i].assign(sum_9_Y = np.where(raw_data[i]["pct_change9"].fillna(0)>=0, 1,0))
    
for i in raw_data.keys():
    raw_data[i] = raw_data[i].dropna(axis=0)

In [6]:
copy_keys = kospi200["종목코드"]
for i in copy_keys:
    if (len(raw_data[i]) < 252):
        del raw_data[i]

In [7]:
kospi200["종목코드"][2]

'282330'

In [8]:
raw_data["006840"]

Open          High           Low         Close  \
Date                                                                 
2012-01-20  14128.200195  14128.200195  13620.799805  13855.000000   
2012-01-25  14011.099609  14011.099609  13659.799805  13855.000000   
2012-01-26  13894.000000  14108.700195  13894.000000  14011.099609   
2012-01-27  14011.099609  14011.099609  13737.900391  13855.000000   
2012-01-30  13913.500000  13972.099609  13484.200195  13503.700195   
2012-01-31  13581.799805  13933.000000  13425.700195  13894.000000   
2012-02-01  14011.099609  14284.299805  13894.000000  14011.099609   
2012-02-02  14225.700195  14498.900391  14167.200195  14284.299805   
2012-02-03  14420.900391  14420.900391  13952.599609  14069.599609   
2012-02-06  14225.700195  14245.299805  13718.400391  13796.400391   
2012-02-07  13855.000000  13894.000000  13581.799805  13718.400391   
2012-02-08  13776.900391  14616.000000  13640.299805  14616.000000   
2012-02-09  14830.700195  14850.200195  14401.400391  14674.599609   
2012-02-10  14830.700195  14889.200195  14284.299805  14303.799805   
2012-02-13  14440.400391  14772.099609  14362.299805  14733.099609   
2012-02-14  14889.200195  14889.200195  14498.900391  14733.099609   
2012-02-15  14830.700195  14830.700195  14420.900391  14635.500000   
2012-02-16  14557.500000  14752.599609  14264.799805  14401.400391   
2012-02-17  14498.900391  14596.500000  14245.299805  14284.299805   
2012-02-20  14459.900391  14557.500000  13991.599609  13991.599609   
2012-02-21  14069.599609  14108.700195  13328.099609  13620.799805   
2012-02-22  13620.799805  13620.799805  13211.000000  13503.700195   
2012-02-23  13484.200195  13679.400391  13269.599609  13523.200195   
2012-02-24  13367.099609  13913.500000  13367.099609  13679.400391   
2012-02-27  13679.400391  13679.400391  13191.500000  13211.000000   
2012-02-28  13211.000000  13347.599609  12957.299805  13347.599609   
2012-02-29  13523.200195  13581.799805  13328.099609  13581.799805   
2012-03-02  13659.799805  13718.400391  13503.700195  13679.400391   
2012-03-05  13757.400391  13757.400391  13425.700195  13464.700195   
2012-03-06  13386.599609  13406.200195  13093.900391  13191.500000   
...                  ...           ...           ...           ...   
2018-06-21  77600.000000  80800.000000  77600.000000  79000.000000   
2018-06-22  79000.000000  79400.000000  74700.000000  76500.000000   
2018-06-25  76600.000000  76600.000000  73800.000000  75500.000000   
2018-06-26  75000.000000  77000.000000  74200.000000  76200.000000   
2018-06-27  75500.000000  76500.000000  74500.000000  74900.000000   
2018-06-28  75300.000000  75800.000000  74300.000000  75300.000000   
2018-06-29  75300.000000  75700.000000  73000.000000  75400.000000   
2018-07-02  75600.000000  75600.000000  70000.000000  70200.000000   
2018-07-03  70800.000000  72000.000000  70100.000000  70100.000000   
2018-07-04  69400.000000  71000.000000  69200.000000  70600.000000   
2018-07-05  70300.000000  71400.000000  70300.000000  71000.000000   
2018-07-06  71000.000000  73000.000000  71000.000000  73000.000000   
2018-07-09  72500.000000  73200.000000  70800.000000  72900.000000   
2018-07-10  72900.000000  75500.000000  72900.000000  74400.000000   
2018-07-11  74000.000000  74600.000000  72400.000000  73200.000000   
2018-07-12  74000.000000  75800.000000  73300.000000  75000.000000   
2018-07-13  75500.000000  77700.000000  75500.000000  77700.000000   
2018-07-16  77900.000000  79400.000000  77200.000000  79000.000000   
2018-07-17  79300.000000  80400.000000  78200.000000  80300.000000   
2018-07-18  80100.000000  80100.000000  77400.000000  78300.000000   
2018-07-19  78100.000000  78100.000000  75100.000000  75800.000000   
2018-07-20  75700.000000  77100.000000  75200.000000  76400.000000   
2018-07-23  77000.000000  77000.000000  74300.000000  74900.000000   
2018-07-24  75400.000000  76400.000000  73300.000000  75800.000000   
2018-07-25  76700.000000  7980

### Train, Test Split

In [9]:
train1 = {}
train_idx1 = {}
test1 = {}
test_idx1 = {}
for code, df in raw_data.items():
    train1[code] = df.iloc[:int(len(df)*0.8)][["Open","ma_5", "ma_10"]]
    train_idx1[code] = df.iloc[:int(len(df)*0.8)]["sum_9_Y"]
    test1[code] = df.iloc[int(len(df)*0.8):][["Open","ma_5", "ma_10"]]
    test_idx1[code] = df.iloc[int(len(df)*0.8):]["sum_9_Y"]

In [10]:
train2 = {}
train_idx2 = {}
test2 = {}
test_idx2 = {}
for code, df in raw_data.items():
    train2[code] = df.iloc[:int(len(df)*0.8)][["Open","ma_5", "ma_15"]]
    train_idx2[code] = df.iloc[:int(len(df)*0.8)]["sum_9_Y"]
    test2[code] = df.iloc[int(len(df)*0.8):][["Open","ma_5", "ma_15"]]
    test_idx2[code] = df.iloc[int(len(df)*0.8):]["sum_9_Y"]

In [11]:
train3 = {}
train_idx3 = {}
test3 = {}
test_idx3 = {}
for code, df in raw_data.items():
    train3[code] = df.iloc[:int(len(df)*0.8)][["Open","ma_10", "ma_15"]]
    train_idx3[code] = df.iloc[:int(len(df)*0.8)]["sum_9_Y"]
    test3[code] = df.iloc[int(len(df)*0.8):][["Open","ma_10", "ma_15"]]
    test_idx3[code] = df.iloc[int(len(df)*0.8):]["sum_9_Y"]

In [12]:
train4 = {}
train_idx4 = {}
test4 = {}
test_idx4 = {}
for code, df in raw_data.items():
    train4[code] = df.iloc[:int(len(df)*0.8)][["Open","ma_5", "ma_10","ma_15"]]
    train_idx4[code] = df.iloc[:int(len(df)*0.8)]["sum_9_Y"]
    test4[code] = df.iloc[int(len(df)*0.8):][["Open","ma_5", "ma_10","ma_15"]]
    test_idx4[code] = df.iloc[int(len(df)*0.8):]["sum_9_Y"]

In [13]:
train5 = {}
train_idx5 = {}
test5 = {}
test_idx5 = {}
for code, df in raw_data.items():
    train5[code] = df.iloc[:int(len(df)*0.8)][["Open","Volume","ma_5", "ma_10","ma_15"]]
    train_idx5[code] = df.iloc[:int(len(df)*0.8)]["sum_9_Y"]
    test5[code] = df.iloc[int(len(df)*0.8):][["Open","Volume","ma_5", "ma_10","ma_15"]]
    test_idx5[code] = df.iloc[int(len(df)*0.8):]["sum_9_Y"]

In [14]:
train6 = {}
train_idx6 = {}
test6 = {}
test_idx6 = {}
for code, df in raw_data.items():
    train6[code] = df.iloc[:int(len(df)*0.8)][["Open","ma_3","ma_5", "ma_10"]]
    train_idx6[code] = df.iloc[:int(len(df)*0.8)]["sum_9_Y"]
    test6[code] = df.iloc[int(len(df)*0.8):][["Open","ma_3","ma_5", "ma_10"]]
    test_idx6[code] = df.iloc[int(len(df)*0.8):]["sum_9_Y"]

### Scale 조정

In [15]:
from sklearn.preprocessing import StandardScaler

std_scaler = StandardScaler()
li1 = ["Open","ma_5", "ma_10"]
for code in raw_data.keys():
    train1[code][li1] = std_scaler.fit_transform(train1[code][li1].values)
    test1[code][li1] = std_scaler.fit_transform(test1[code][li1].values)

In [16]:
from sklearn.preprocessing import StandardScaler

std_scaler = StandardScaler()
li2 = ["Open","ma_5", "ma_15"]
for code in raw_data.keys():
    train2[code][li2] = std_scaler.fit_transform(train2[code][li2].values)
    test2[code][li2] = std_scaler.fit_transform(test2[code][li2].values)

In [17]:
from sklearn.preprocessing import StandardScaler

std_scaler = StandardScaler()
li3 = ["Open","ma_10", "ma_15"]
for code in raw_data.keys():
    train3[code][li3] = std_scaler.fit_transform(train3[code][li3].values)
    test3[code][li3] = std_scaler.fit_transform(test3[code][li3].values)

In [18]:
from sklearn.preprocessing import StandardScaler

std_scaler = StandardScaler()
li4 = ["Open","ma_5", "ma_10","ma_15"]
for code in raw_data.keys():
    train4[code][li4] = std_scaler.fit_transform(train4[code][li4].values)
    test4[code][li4] = std_scaler.fit_transform(test4[code][li4].values)

In [19]:
from sklearn.preprocessing import StandardScaler

std_scaler = StandardScaler()
li5 = ["Open","Volume","ma_5", "ma_10","ma_15"]
for code in raw_data.keys():
    train5[code][li5] = std_scaler.fit_transform(train5[code][li5].values)
    test5[code][li5] = std_scaler.fit_transform(test5[code][li5].values)

In [20]:
from sklearn.preprocessing import StandardScaler

std_scaler = StandardScaler()
li6 = ["Open","ma_3","ma_5", "ma_10"]
for code in raw_data.keys():
    train6[code][li6] = std_scaler.fit_transform(train6[code][li6].values)
    test6[code][li6] = std_scaler.fit_transform(test6[code][li6].values)

### model 호출

In [21]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

log_clf = LogisticRegression()
DTC_clf = DecisionTreeClassifier()
svm_clf = SVC()

### Predict

In [22]:
score_list1 = pd.DataFrame(columns=('LR','DT','SVM'))

for i in raw_data.keys():
    a1=[]
    X_train1 = train1[i].values
    y_train1 = train_idx1[i].values
    X_test1 = test1[i].values
    y_test1 = test_idx1[i].values
    for clf in (log_clf, DTC_clf, svm_clf):
        clf.fit(X_train1, y_train1)
        y_pred1 = clf.predict(X_test1)
        a1.append(accuracy_score(y_test1, y_pred1))
    score_list1.loc[i] = a1

In [23]:
score_list2 = pd.DataFrame(columns=('LR','DT','SVM'))

for i in raw_data.keys():
    a2=[]
    X_train2 = train2[i].values
    y_train2 = train_idx2[i].values
    X_test2 = test2[i].values
    y_test2 = test_idx2[i].values
    for clf in (log_clf, DTC_clf, svm_clf):
        clf.fit(X_train2, y_train2)
        y_pred2 = clf.predict(X_test2)
        a2.append(accuracy_score(y_test2, y_pred2))
    score_list2.loc[i] = a2

In [24]:
score_list3 = pd.DataFrame(columns=('LR','DT','SVM'))

for i in raw_data.keys():
    a3=[]
    X_train3 = train3[i].values
    y_train3 = train_idx3[i].values
    X_test3 = test3[i].values
    y_test3 = test_idx3[i].values
    for clf in (log_clf, DTC_clf, svm_clf):
        clf.fit(X_train3, y_train3)
        y_pred3 = clf.predict(X_test3)
        a3.append(accuracy_score(y_test3, y_pred3))
    score_list3.loc[i] = a3

In [25]:
score_list4 = pd.DataFrame(columns=('LR','DT','SVM'))

for i in raw_data.keys():
    a4=[]
    X_train4 = train4[i].values
    y_train4 = train_idx4[i].values
    X_test4 = test4[i].values
    y_test4 = test_idx4[i].values
    for clf in (log_clf, DTC_clf, svm_clf):
        clf.fit(X_train4, y_train4)
        y_pred4 = clf.predict(X_test4)
        a4.append(accuracy_score(y_test4, y_pred4))
    score_list4.loc[i] = a4

In [26]:
score_list5 = pd.DataFrame(columns=('LR','DT','SVM'))

for i in raw_data.keys():
    a5=[]
    X_train5 = train5[i].values
    y_train5 = train_idx5[i].values
    X_test5 = test5[i].values
    y_test5 = test_idx5[i].values
    for clf in (log_clf, DTC_clf, svm_clf):
        clf.fit(X_train5, y_train5)
        y_pred5 = clf.predict(X_test5)
        a5.append(accuracy_score(y_test5, y_pred5))
    score_list5.loc[i] = a5

In [27]:
from sklearn.metrics import accuracy_score

score_list6 = pd.DataFrame(columns=('LR','DT','SVM'))


for i in raw_data.keys():
    a6=[]
    X_train6 = train6[i].values
    y_train6 = train_idx6[i].values
    X_test6 = test6[i].values
    y_test6 = test_idx6[i].values
    for clf in (log_clf, DTC_clf, svm_clf):
        clf.fit(X_train6, y_train6)
        y_pred6 = clf.predict(X_test6)
        a6.append(accuracy_score(y_test6, y_pred6))
    score_list6.loc[i] = a6

In [28]:
# Y : 5일평균주가 등락 

print(score_list1.describe()) #open, ma5, ma10 // 성능이 가장 좋음
print(score_list2.describe()) #open, ma5, ma15
print(score_list3.describe()) #open, ma10, ma15
print(score_list4.describe()) # open, ma5, ma10, ma15
print(score_list5.describe()) #open, ,volume, ma5, ma10, ma15
print(score_list6.describe()) #open, ma3, ma5, ma10

               LR          DT         SVM
count  197.000000  197.000000  197.000000
mean     0.804962    0.742088    0.767048
std      0.054971    0.049653    0.078292
min      0.419255    0.470588    0.411765
25%      0.788820    0.717391    0.748447
50%      0.813665    0.748447    0.788820
75%      0.832298    0.779503    0.816176
max      0.897143    0.826087    0.900621
               LR          DT         SVM
count  197.000000  197.000000  197.000000
mean     0.829972    0.756498    0.810072
std      0.044877    0.049771    0.063201
min      0.475490    0.563725    0.421569
25%      0.810559    0.726708    0.795031
50%      0.835404    0.763975    0.822981
75%      0.857143    0.795031    0.847826
max      0.913043    0.854037    0.916149
               LR          DT         SVM
count  197.000000  197.000000  197.000000
mean     0.807930    0.740378    0.790367
std      0.051970    0.047334    0.062879
min      0.400621    0.578431    0.421569
25%      0.791925    0.714286    0

In [29]:
# Y : 7일평균주가 등락 

print(score_list1.describe()) #open, ma5, ma10
print(score_list2.describe()) #open, ma5, ma15
print(score_list3.describe()) #open, ma10, ma15
print(score_list4.describe()) # open, ma5, ma10, ma15
print(score_list5.describe()) #open, ,volume, ma5, ma10, ma15
print(score_list6.describe()) #open, ma3, ma5, ma10 // 성능이 가장 좋음

               LR          DT         SVM
count  197.000000  197.000000  197.000000
mean     0.804962    0.742088    0.767048
std      0.054971    0.049653    0.078292
min      0.419255    0.470588    0.411765
25%      0.788820    0.717391    0.748447
50%      0.813665    0.748447    0.788820
75%      0.832298    0.779503    0.816176
max      0.897143    0.826087    0.900621
               LR          DT         SVM
count  197.000000  197.000000  197.000000
mean     0.829972    0.756498    0.810072
std      0.044877    0.049771    0.063201
min      0.475490    0.563725    0.421569
25%      0.810559    0.726708    0.795031
50%      0.835404    0.763975    0.822981
75%      0.857143    0.795031    0.847826
max      0.913043    0.854037    0.916149
               LR          DT         SVM
count  197.000000  197.000000  197.000000
mean     0.807930    0.740378    0.790367
std      0.051970    0.047334    0.062879
min      0.400621    0.578431    0.421569
25%      0.791925    0.714286    0

In [30]:
# Y : 9일평균주가 등락

print(score_list1.describe()) #open, ma5, ma10
print(score_list2.describe()) #open, ma5, ma15 // SVM 성능 가장 좋음
print(score_list3.describe()) #open, ma10, ma15
print(score_list4.describe()) # open, ma5, ma10, ma15 // LR,DT 성능 가장 좋음
print(score_list5.describe()) #open, ,volume, ma5, ma10, ma15
print(score_list6.describe()) #open, ma3, ma5, ma10

               LR          DT         SVM
count  197.000000  197.000000  197.000000
mean     0.804962    0.742088    0.767048
std      0.054971    0.049653    0.078292
min      0.419255    0.470588    0.411765
25%      0.788820    0.717391    0.748447
50%      0.813665    0.748447    0.788820
75%      0.832298    0.779503    0.816176
max      0.897143    0.826087    0.900621
               LR          DT         SVM
count  197.000000  197.000000  197.000000
mean     0.829972    0.756498    0.810072
std      0.044877    0.049771    0.063201
min      0.475490    0.563725    0.421569
25%      0.810559    0.726708    0.795031
50%      0.835404    0.763975    0.822981
75%      0.857143    0.795031    0.847826
max      0.913043    0.854037    0.916149
               LR          DT         SVM
count  197.000000  197.000000  197.000000
mean     0.807930    0.740378    0.790367
std      0.051970    0.047334    0.062879
min      0.400621    0.578431    0.421569
25%      0.791925    0.714286    0

## Y : 7일평균주가 등락, Y : 9일평균주가 등락 두 경우 모두 LR는 83% 였지만 DT는 77%,75% SVM의 경우 각각 79%, 81% 이 나왔다.